### PR TITLE
feat(Table): table settings component added

### PR DESCRIPTION
**Pull request by TaPo4eK111291**

### CODE REVIEW

File: src/components/TableSettingsColumn/TableSettingsColumn.tsx [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    };
+
+    const {attributes, listeners, setNodeRef, transform, transition, isDragging} = useSortable({
+        id: column.id,
+    });
+
+    const style = {
+        transform: CSS.Transform.toString(transform),
+        transition,
+    };
```
Comment (beliarh) [inline src/components/TableSettingsColumn/TableSettingsColumn.tsx:72]: There must be a bug with `transform` style:

https://github.com/user-attachments/assets/db194aa0-3cfa-4b17-8d06-71096bb8d2bf

**Review by beliarh — CHANGES_REQUESTED** [review]
1. Is it possible to prevent some columns from hiding? (by disabling checkbox for example)
2. Parent column's checkbox is not clickable for some reason
3. What about supporting the column pinning feature?

File: src/utils/addComponentKeysets.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import type {KeysData} from '@gravity-ui/i18n';
+
+import {i18n} from '../i18n';
+
+import type {Lang} from './configure';
+
+export function addComponentKeysets<T extends KeysData>(data: Record<Lang, T>, keysetName: string) {
```
[the remark this answers is not in the document]
Comment (TaPo4eK111291) [reply src/utils/addComponentKeysets.ts]: Reused addComponentsKeysets from @gravity-ui/uikit/i18n

File: src/components/TableSettingsColumn/TableSettingsColumn.tsx [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    return (
+        <div
+            key={column.id}
+            className={b('column')}
```
[the remark this answers is not in the document]
Comment (TaPo4eK111291) [reply src/components/TableSettingsColumn/TableSettingsColumn.tsx]: Don't want to mix styles of table-settings with table-settings-column, but I left the isolated component and corrected the classes to:
gt-table-settings-column__column => gt-table-settings-column
gt-table-settings-column__column-content => gt-table-settings-column__content

File: src/components/TableSettings/TableSettings.tsx [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+    const {orderedItems, handleDragEnd} = useOrderedItems(
+        filteredColumns,
+        orderState,
+        setOrderState,
+    );
+
+    return (
+        <React.Fragment>
+            <Popup open={open} onClose={() => setOpen(false)} anchorRef={anchorRef}>
```
[the remark this answers is not in the document]
Comment (TaPo4eK111291) [reply src/components/TableSettings/TableSettings.tsx]: Set position like uikit table.

File: package.json [hunk 5]
```diff
@@ -43,6 +43,7 @@
   },
   "dependencies": {
     "@bem-react/classname": "^1.6.0",
+    "@gravity-ui/i18n": "^1.6.0",
```
[the remark this answers is not in the document]
Comment (TaPo4eK111291) [reply package.json]: Moved

File: src/components/TableSettings/TableSettings.tsx [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    table: Table<TData>;
+    column?: Column<TData, unknown>;
+}
+
+export const TableSettings = <TData extends unknown>({table, column}: Props<TData>) => {
+    const anchorRef = React.useRef<HTMLButtonElement>(null);
+    const [open, setOpen] = React.useState<boolean>(false);
+    const columns = table.getAllColumns();
+    const filteredColumns = column
+        ? columns.filter((otherColumn) => otherColumn.id !== column.id)
```
[the remark this answers is not in the document]
Comment (TaPo4eK111291) [reply src/components/TableSettings/TableSettings.tsx]: Wrapped computed properties by useMemo.

File: src/components/TableSettings/__stories__/stories/TableSettingsColumnStory.tsx [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+type Item = {
+    operatorName: string;
+    contactGroupName: string;
+    contactName: string;
+    contactPhone: string;
+    contactGroupType: string;
+    contactPriority: string;
+};
+
+const data = [
```
[the remark this answers is not in the document]
Comment (TaPo4eK111291) [reply src/components/TableSettings/__stories__/stories/TableSettingsColumnStory.tsx]: Added

File: src/components/TableSettings/__stories__/stories/TableSettingsColumnStory.tsx [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        contactGroupType: 'Admin support',
+        contactPriority: '1',
+    },
+];
+
+const columns: ColumnDef<Item>[] = [
+    {
+        id: 'contact_group_group',
+        header: 'Contact Group',
+        size: 100,
```
[the remark this answers is not in the document]
Comment (TaPo4eK111291) [reply src/components/TableSettings/__stories__/stories/TableSettingsColumnStory.tsx]: Looks like, parent column size ignores by Table, if all children has sizes. Removed sized from parent columns.

File: src/components/TableSettingsColumn/TableSettingsColumn.tsx (hunk 1 above)
Comment (TaPo4eK111291) [reply src/components/TableSettingsColumn/TableSettingsColumn.tsx:79]: Changed layout and set override for scaleY to 1, now it's looks much better, thx.

**Comment by TaPo4eK111291** [comment]
> * Is it possible to prevent some columns from hiding? (by disabling checkbox for example)
> * Parent column's checkbox is not clickable for some reason
> * What about supporting the column pinning feature?

1. Added reaction to enableHiding prop in columns config. Now u can set it to false, for prevent toggling.
2. Fixed.
3. Right now we don't have mockups for pinning and disabling other behaviours, we'll add them separately.

**Comment by Raubzeug** [comment]
Name of group name should be bold.
Screenshot 2024-09-24 at 11 36 58

**Comment by Raubzeug** [comment]
Drag handler is not centered:
Screenshot 2024-09-24 at 11 38 09

**Comment by Raubzeug** [comment]
Column with settings handler should always be visible (pinned on the end of the flow (depends on rtl or ltr)).

**Review by Raubzeug — CHANGES_REQUESTED** [review]

**Comment by TaPo4eK111291** [comment]
> Name of group name should be bold. Screenshot 2024-09-24 at 11 36 58

Fixed

File: src/components/TableSettings/TableSettings.tsx [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+    const renderColumns = (renderedColumns: Column<TData, unknown>[]) => {
+        return renderedColumns.map((innerColumn, index) => {
+            const children = renderColumns(innerColumn.columns);
+            const header = headersById[innerColumn.id];
+            const rootNode = innerColumn.depth === 0;
+            const lastInGroup = index === renderedColumns.length - 1;
+
+            return (
+                <React.Fragment key={innerColumn.id}>
```
[the remark this answers is not in the document]
Comment (TaPo4eK111291) [reply src/components/TableSettings/TableSettings.tsx:70]: Removed React.Fragment

File: src/components/TableSettings/TableSettings.scss [hunk 10]
```diff
@@ -0,0 +1,13 @@
+@use '../variables';
+
+$block: '.#{variables.$ns}table-settings';
+
+#{$block} {
+    &__popover-content {
+        padding: 4px;
```
[the remark this answers is not in the document]
Comment (TaPo4eK111291) [reply src/components/TableSettings/TableSettings.scss:7]: Fixed.

File: src/components/TableSettingsColumn/TableSettingsColumn.tsx [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+    });
+
+    if (transform) transform.scaleY = 1;
+    const style = {
+        transform: CSS.Transform.toString(transform),
+        transition,
+    };
+
+    return (
+        <div style={style} ref={setNodeRef}>
```
[the remark this answers is not in the document]
Comment (TaPo4eK111291) [reply src/components/TableSettingsColumn/TableSettingsColumn.tsx:90]: Fixed

File: src/components/TableSettings/TableSettings.tsx [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+                        showDivider={rootNode && !lastInGroup}
+                    >
+                        {children}
+                    </TableSettingsColumn>
+                </React.Fragment>
+            );
+        });
+    };
+
+    const applyNewSettings = () => {
```
[the remark this answers is not in the document]
Comment (TaPo4eK111291) [reply src/components/TableSettings/TableSettings.tsx:88]: Added onSettingsApply callback for TableSettings component

File: src/constants/settingsColumn.tsx [hunk 13]
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+
+import type {ColumnDef} from '@tanstack/react-table';
+
+import {TableSettings} from '../components/TableSettings/TableSettings';
+import type {TableSettingsOptions} from '../components/TableSettings/TableSettings';
+
+export const settingsColumn: ColumnDef<unknown> = {
```
[the remark this answers is not in the document]
Comment (TaPo4eK111291) [reply src/constants/settingsColumn.tsx:11]: Removed

File: src/components/TableSettings/TableSettings.utils.ts [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+            ...item,
+            columns: orderedChildren?.map((child) => orderItem(child, state)) ?? [],
+        };
+    };
+
+    const orderedItems =
+        orderItem({id: 'root', columns: items} as Column<TData, unknown>, orderState).columns ?? [];
+
+    const handleDragEnd = ({active, over}: DragEndEvent) => {
+        const activeContainer = findContainer(orderState, active.id as string);
```
[the remark this answers is not in the document]
Comment (TaPo4eK111291) [reply src/components/TableSettings/TableSettings.utils.ts:33]: Removed all type casting

File: src/components/TableSettings/TableSettings.tsx [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+            (acc, header) => {
+                const result = {...acc};
+                result[header.column.id] = header;
+                return acc;
+            },
+            {} as Record<string, Header<TData, unknown>>,
+        );
+    }, [headers]);
+
+    const [visibilityState, setVisibilityState] = React.useState(table.getState().columnVisibility);
```
[the remark this answers is not in the document]
Comment (TaPo4eK111291) [reply src/components/TableSettings/TableSettings.tsx:59]: Fixed

File: src/utils/configure.ts [hunk 16]
```diff
@@ -0,0 +1,33 @@
+export enum Lang {
```
[the remark this answers is not in the document]
Comment (TaPo4eK111291) [reply src/utils/configure.ts]: Removed all copypaste, thx.

**Review by Raubzeug — DISMISSED** [review]

File: src/components/TableSettings/TableSettings.utils.ts [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+};
+
+export const useOrderedItems = <TData extends unknown>(
+    items: Column<TData>[],
+    orderState: Record<string, string[]>,
+    setOrderState: React.Dispatch<React.SetStateAction<Record<string, string[]>>>,
+) => {
+    const [activeDepth, setActiveDepth] = React.useState<number | undefined>();
+
+    const dephMap = React.useMemo(() => {
```
Comment (beliarh) [inline src/components/TableSettings/TableSettings.utils.ts:37]: `dephMap` => `depthMap`

File: src/components/TableSettingsColumn/TableSettingsColumn.utils.ts [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+export const getIsVisible = <TData extends unknown>(
+    column: Column<TData>,
+    visibilityState: VisibilityState,
+): boolean => {
+    if (column.columns.length)
+        return column.columns.some((innerColumn) => getIsVisible(innerColumn, visibilityState));
+    return visibilityState[column.id] ?? true;
+};
+
+export const isEnabledHidding = <TData extends unknown>(column: Column<TData>): boolean =>
```
Comment (beliarh) [inline src/components/TableSettingsColumn/TableSettingsColumn.utils.ts:12]: `isEnabledHidding` => `isEnabledHiding`

**Review by beliarh — CHANGES_REQUESTED** [review]

File: src/components/TableSettings/TableSettings.utils.ts (hunk 17 above)
Comment (TaPo4eK111291) [reply src/components/TableSettings/TableSettings.utils.ts]: Fixed

File: src/components/TableSettingsColumn/TableSettingsColumn.utils.ts (hunk 18 above)
Comment (TaPo4eK111291) [reply src/components/TableSettingsColumn/TableSettingsColumn.utils.ts]: Fixed

**Review by beliarh — APPROVED** [review]

**Pull request merged**